src/box/memtx_tx.c

```diff
@@ -2996,13 +2961,87 @@ memtx_tx_delete_gap(struct gap_item_base *item)
 	memtx_tx_mempool_free(item->txn, pool, item);
 }
 
+/**
+ * Abort all concurrent readers of the old schema of the space.
+ */
 void
-memtx_tx_invalidate_space(struct space *space, struct txn *active_txn)
+memtx_tx_abort_space_schema_readers(struct space *space, struct txn *ddl_owner)
 {
+	/* Abort all story and gap-with-sucessor readers. */
+	struct memtx_story *story;
+	rlist_foreach_entry(story, &space->memtx_stories, in_space_stories) {
+		struct tx_read_tracker *tracker, *tmp;
+		rlist_foreach_entry_safe(tracker, &story->reader_list,
+					 in_reader_list, tmp) {
+			if (tracker->reader != ddl_owner)
+				memtx_tx_abort_with_conflict(tracker->reader);
+		}
+
+		for (uint32_t i = 0; i < story->index_count; i++) {
+			if (story->link[i].newer_story != NULL)
+				continue;
+			struct gap_item_base *item, *tmp;
+			rlist_foreach_entry_safe(item,
+						 &story->link[i].read_gaps,
+						 in_read_gaps, tmp) {
+				if (item->txn != ddl_owner)
+					memtx_tx_abort_with_conflict(item->txn);
+			}
+		}
+	}
+
+	/* Abort all gap-without-successor readers. */
+	for (uint32_t i = 0; i < space->index_count; i++) {
+		struct index *index = space->index[i];
+		struct gap_item_base *item, *tmp;
+		rlist_foreach_entry_safe(item, &index->read_gaps,
+					 in_read_gaps, tmp) {
+			if (item->txn != ddl_owner)
+				memtx_tx_abort_with_conflict(item->txn);
+		}
+	}
+
+	/*
+	 * Iterate over all transactions in order to:
+	 * 1. Abort all writers.
+	 * 2. Abort all point hole readers.
+	 */
+	struct txn *txn;
+	rlist_foreach_entry(txn, &txm.all_txs, in_all_txs) {
+		if (txn == ddl_owner)
+			continue;
+		struct txn_stmt *stmt;
+		stailq_foreach_entry(stmt, &txn->stmts, next) {
+			if (stmt->space == space) {
+				memtx_tx_abort_with_conflict(txn);
+				break;
+			}
+		}
+		struct point_hole_item *hole_item;
+		rlist_foreach_entry(hole_item, &txn->point_holes_list,
+				    in_point_holes_list) {
+			if (space->def->id == hole_item->space_id) {
+				memtx_tx_abort_with_conflict(txn);
```

Comment:
We must abort transactions writing to the space because otherwise, they'd crash on commit (`txn_stmt` stores a reference to a space). Do we really need to abort readers? We don't do that in Vinyl.

Reply:
1. Some trackers store keys that are compared to tuples with `key_def`. It's dangerous to compare old keys with a new `key_def` - for example, an `!is_nullable` optimization could be enabled after schema change - old nullable key could lead to a crash in this case. By the way, isn't vinyl affected by this problem too? Doesn't it store keys in read sets of transactions?
2. On space invalidation we remove all the stories and insert tuples according to what `active_ddl` observes. If some RO transactions were not aborted, we would need to think if we don't violate their isolation.

Reply:
> 1. Some trackers store keys that are compared to tuples with `key_def`. It's dangerous to compare old keys with a new `key_def` - for example, an `!is_nullable` optimization could be enabled after schema change - old nullable key could lead to a crash in this case. By the way, isn't vinyl affected by this problem too? Doesn't it store keys in read sets of transactions?

Vinyl read sets are stored in indexes (`vy_lsm::read_set`), i.e. a read set is dropped along with the index and its key definition. Doesn't it work the same way in memtx?

> 2. On space invalidation we remove all the stories and insert tuples according to what `active_ddl` observes. If some RO transactions were not aborted, we would need to think if we don't violate their isolation.

AFAIU letting read-only transactions proceed wouldn't break transaction serializability - we could assume that they were executed before the DROP/ALTER statement. The code would be so much simpler then and consistent with Vinyl.

Reply:
Discussed f2f - decided to abort all RO transactions in memtx.